### PR TITLE
IFU release v2.6

### PR DESCRIPTION
**Pull request by wangye805**
# Description

upstream release_v2.6 (with commit c90a7207) IFU based on dev commit (669b556b)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Infra/Build change
- [x] Code refactoring

## Changes
Resolve several conflicts

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/pytorch/attention/test_attention.py [hunk 1]
```diff
@@ -525,40 +346,29 @@ def test_dpa_checkpoint(dtype, model_configs, model):
...
-        8, 16, 16, 256, 1, 2048, 0.0, "no_mask", "no_bias", head_dim_v=128
-    ),  # inference
-    "mla_3_2": ModelConfig(
-        8, 16, 16, 192, 1, 2048, 0.0, "no_mask", "no_bias", head_dim_v=128
-    ),  # inference
+    "mla_3_0": ModelConfig(8, 1, 16, 128, max_seqlen_kv=2048, head_dim_v=64),  # inference
+    "mla_3_1": ModelConfig(8, 1, 16, 256, max_seqlen_kv=2048, head_dim_v=128),  # inference
+    "mla_3_2": ModelConfig(8, 1, 16, 192, max_seqlen_kv=2048, head_dim_v=128),  # inference
+}
+if IS_HIP_EXTENSION:
```
Comment (ipanfilo) [inline tests/pytorch/attention/test_attention.py]: With adding mla_1_2 and mla_2_2 configs, mla_4_* are not needed anymore

File: tests/pytorch/test_numerics.py [hunk 2]
```diff
@@ -142,6 +129,26 @@ def __init__(self, hidden_size, eps, num_attention_heads, embed, num_layers, seq
 ]
 
 
+def is_fused_attn_available(
+    config: ModelConfig, dtype: torch.dtype, qkv_layout="bshd_bshd_bshd", is_training=True
+):
+    # backup the NVTE_FUSED_ATTN_* envs
+    if IS_HIP_EXTENSION:
+        from utils import EnvVarCleaner
```
Comment (ipanfilo) [inline tests/pytorch/test_numerics.py:137]: Since EnvVarCleaner is moved to utils, its declaration at line 220 should be removed as well

File: tests/pytorch/attention/test_kv_cache.py
Comment (ipanfilo) [inline tests/pytorch/attention/test_kv_cache.py:386]: Have you ever got test_kv_cache failure with FlashAttention and default tolerance?

File: tests/pytorch/distributed/test_sanity.py
Comment (ipanfilo) [inline tests/pytorch/distributed/test_sanity.py:1]: Do we need it in CI script?

File: transformer_engine/common/normalization/common.h [hunk 3]
```diff
@@ -29,6 +29,7 @@
 #ifndef __HIP_PLATFORM_AMD__
 #include "../cudnn_utils.h"
 #else
+#include "../util/ptx.cuh"
```
Comment (ipanfilo) [inline transformer_engine/common/normalization/common.h]: It should rather be included to "rocm_cast_kernel.cuh where it is actually used"

File: transformer_engine/common/util/rocm_dequantize_kernels.cuh [hunk 4]
```diff
@@ -102,7 +102,7 @@ __global__ void __launch_bounds__(THREADS_PER_CHUNK)
 
     const int scale_idx = scale_offset_Y * scales_stride + scale_offset_X;
     const e8m0_t biased_exponent = scales_ptr[scale_idx];
-    const float block_scale = exp2f(static_cast<float>(biased_exponent) - FP32_EXPONENT_BIAS);
+    const float block_scale = exp2f(static_cast<float>(biased_exponent) - ptx::FP32_EXPONENT_BIAS);
```
Comment (ipanfilo) [inline transformer_engine/common/util/rocm_dequantize_kernels.cuh]: Its better to be just ptx::exp2f(biased_exponent)

File: benchmarks/attention/benchmark_attention_rocm.py [hunk 5]
```diff
@@ -46,12 +50,12 @@
...
-    "test_1": ModelConfig(2, 16, 16, 128, 2048, 2048, 0.0, "causal", "no_bias"),  # longer seq, mask
-    "test_2": ModelConfig(2, 16, 16, 128, 2048, 2048, 0.0, "causal", "post_scale_bias"),  # bias
-    "test_3": ModelConfig(2, 32, 4, 128, 8192, 8192, 0.0, "causal", "no_bias"),  # GQA
-    "test_4": ModelConfig(2, 128, 8, 128, 8192, 8192, 0.0, "causal_bottom_right", "no_bias")
+    #   test:             b,  sq, h, d
+    "test_0": ModelConfig(2, 512, 16, 64),  # short seq
+    "test_1": ModelConfig(2, 2048, 16, 128, attn_mask_type="causal"),  # longer seq, mask
+    "test_2": ModelConfig(2, 2048, 16, 128, attn_mask_type="causal", attn_bias_type="post_scale_bias"),  # bias
+    "test_3": ModelConfig(2, 8192, 32, 128, num_gqa_groups=8, attn_mask_type="causal"),  # GQA
+    "test_4": ModelConfig(2, 8192, 128, 128, num_gqa_groups=16, attn_mask_type="causal_bottom_right")
```
Comment (ipanfilo) [inline benchmarks/attention/benchmark_attention_rocm.py]: Why num_gqa_groups is changed from original for test_3 and test_4?

File: tests/pytorch/attention/test_kv_cache.py
Comment (wangye805) [reply tests/pytorch/attention/test_kv_cache.py:386]: Those special tolerance setting was from NV upstream. I didn't touch this file entirely and not sure whether test_kv_cache fails with flash-attn + default tol.

Comment (wangye805) [reply tests/pytorch/attention/test_kv_cache.py:386]: Just checked our pytorch ci script, we didn't enable this test series before

File: tests/pytorch/attention/test_attention.py (hunk 1 above)
Comment (wangye805) [reply tests/pytorch/attention/test_attention.py]: Addressed. Thanks

File: tests/pytorch/test_numerics.py (hunk 2 above)
Comment (wangye805) [reply tests/pytorch/test_numerics.py:137]: Removed.

File: tests/pytorch/distributed/test_sanity.py
Comment (wangye805) [reply tests/pytorch/distributed/test_sanity.py:1]: Added in

File: transformer_engine/common/normalization/common.h (hunk 3 above)
Comment (wangye805) [reply transformer_engine/common/normalization/common.h]: Moved

File: transformer_engine/common/util/rocm_dequantize_kernels.cuh (hunk 4 above)
Comment (wangye805) [reply transformer_engine/common/util/rocm_dequantize_kernels.cuh]: Addressed. Thanks

File: benchmarks/attention/benchmark_attention_rocm.py (hunk 5 above)
Comment (wangye805) [reply benchmarks/attention/benchmark_attention_rocm.py]: Changed it back. Thanks

File: .github/workflows/rocm-ci.yml [hunk 6]
```diff
@@ -248,10 +248,9 @@ jobs:
           HIP_VISIBLE_DEVICES=2 ci/jax.sh > /workspace/jax_sgpu.log 2>&1 &
           jax_pid=$!; echo JAX test pid $!
           
-          HIP_VISIBLE_DEVICES=3 ci/core.sh > /workspace/core_sgpu.log 2>&1 &
-          core_pid=$!; echo Core test pid $!
+          ci/core.sh > /workspace/core_sgpu.log 2>&1
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci.yml]: Why is it a part of release?

File: tests/pytorch/test_numerics.py (hunk 2 above)
Comment (ipanfilo) [reply tests/pytorch/test_numerics.py:137]: I'm confused how does it work now. EnvVarCleaner is only imported when is_fused_attn_available but it is used in fixture at line 220

**Review by ipanfilo — COMMENTED** [review]
Please update copyright year on files that continue to modify

File: .github/workflows/rocm-ci.yml (hunk 6 above)
Comment (wangye805) [reply .github/workflows/rocm-ci.yml]: This is from cherry-picking https://github.com/ROCm/TransformerEngine/pull/426 from dev branch. We found that sgpu cpp core gtest always use all 8 GPUs which can affect the sm_count pytorch pytests.

File: tests/pytorch/test_numerics.py (hunk 2 above)
Comment (wangye805) [reply tests/pytorch/test_numerics.py:137]: Ah, I forgot to import it from util.

File: .github/workflows/rocm-ci.yml [hunk 7]
```diff
@@ -40,9 +40,13 @@ concurrency:
...
   build_and_test:
-    name: Build and Test on GPU
+    name: Build and Test on GPU (${{ matrix.runner }})
     timeout-minutes: 720
-    runs-on: linux-mi325-8
+    runs-on: ${{ matrix.runner }}
+    strategy:
+      fail-fast: false
+      matrix:
+        runner: [linux-mi325-8, linux-mi355-8]
```
Comment (ipanfilo) [inline .github/workflows/rocm-ci.yml]: It may need cherry-pick of CI runners labels update PR

Comment (wangye805) [reply .github/workflows/rocm-ci.yml]: Done. Thanks

**Review by ipanfilo — APPROVED** [review]

File: qa/L3_pytorch_FA_versions_test/test.sh [hunk 8]
```diff
@@ -18,10 +18,10 @@ sm_arch=`python3 -c "import torch; sm = torch.cuda.get_device_capability(0); pri
 export FLASH_ATTN_CUDA_ARCHS=$sm_arch
 if [ $sm_arch -gt 90 ]
 then
-  FA_versions=(2.8.1)
+  FA_versions=(2.8.3)
```
Comment (wenchenvincent) [inline qa/L3_pytorch_FA_versions_test/test.sh:21]: @ipanfilo @wangye805 Our CI image has not been using FA 2.8.3, right? If so, why do we want to upgrade to 2.8.3?

Comment (ipanfilo) [reply qa/L3_pytorch_FA_versions_test/test.sh:21]: > @ipanfilo @wangye805 Our CI image has not been using FA 2.8.3, right? If so, why do we want to upgrade to 2.8.3?

Our CI image uses 2.8.1 for compatibility with release 2.4. But the plan is to switch to 2.8.3 on the next image (ROCm 7.2.x)

**Review by wenchenvincent — APPROVED** [review]

**Comment by wenchenvincent** [comment]
@wangye805 Have you run level 3 CI?

**Comment by wangye805** [comment]
> @wangye805 Have you run level 3 CI?

Yes it fails the pytorch distributed test_sanity.py again. I'm checking on it.
